File: app/src/main/java/com/example/groxweatherapp/grox/WeatherSuccessAction.java
package com.example.groxweatherapp.grox;

import com.example.groxweatherapp.api.WeatherApiClient.RequestMode;
import com.example.groxweatherapp.model.WeatherModel;
import com.example.groxweatherapp.model.WeatherResponse;
import com.groupon.grox.Action;

import static com.example.groxweatherapp.model.WeatherModel.WeatherModelState.SUCCESS;

public class WeatherSuccessAction implements Action<WeatherModel> {

    private final WeatherResponse weatherResponse;
    @RequestMode private final int requestMode;

    public WeatherSuccessAction(WeatherResponse weatherResponse, @RequestMode int requestMode) {
        this.weatherResponse = weatherResponse;
        this.requestMode = requestMode;
    }

    @Override
    public WeatherModel newState(WeatherModel oldState) {
        return oldState.toBuilder()
            .setState(SUCCESS)
            .setRequestMode(requestMode)
            .setWeather(weatherResponse.weathers)
            .build();
    }
}

File: app/src/main/java/com/example/groxweatherapp/grox/WeatherSuccessAction.kt
package com.example.groxweatherapp.grox

import com.example.groxweatherapp.api.WeatherApiClient.Companion.ForecastMode
import com.example.groxweatherapp.model.WeatherModel
import com.example.groxweatherapp.model.WeatherModel.Companion.WeatherModelState.SUCCESS
import com.example.groxweatherapp.model.WeatherResponse
import com.groupon.grox.Action

class WeatherSuccessAction(private val weatherResponse: WeatherResponse, @ForecastMode private val forecastMode: Int) : Action<WeatherModel> {

    override fun newState(oldState: WeatherModel): WeatherModel {
        return oldState.copy(
                state = SUCCESS,
                forecastMode = forecastMode,
                weatherList = weatherResponse.weathers)
    }
}
